000-mm2-connector-offsets-support.md

```diff
@@ -0,0 +1,447 @@
+However, for each of the MirrorMaker connectors (MirrorSourceConnector, MirrorCheckpointConnector, MirrorHeartbeatConnector) we know the expected structure.
+
+MirrorSourceConnector format:
+```
+{
+  "offsets": [
+    {
+      "partition": {
+        "cluster": "east-kafka",
+        "partition": 0,
+        "topic": "mirrormaker2-cluster-configs"
+      },
+      "offset": {
+        "offset": 0
+      }
+    }
+  ]
+}
+```
+
+MirrorCheckpointConnector format:
+```
+{
+  "offsets": [
+    {
+      "partition": {
+        "partition": 4,
+        "topic": "inventory",
+        "group": "my-group"
+      },
+      "offset": {
+        "offset": 0
+      }
+    }
+  ]
+}
+```
+
+MirrorHeartbeatConnector format:
+```
+{
+  "offsets": [
+    {
+      "partition": {
+        "sourceClusterAlias": "east-kafka",
+        "targetClusterAlias": "west-kafka"
+      },
+      "offset": {
+        "offset": 0
+      }
+    }
+  ]
+}
+```
+
+### New annotations
+
+All three actions will make use of a new annotation called `strimzi.io/connector-offsets`.
+The possible values will be `list`, `alter`, and `reset`.
+
+A second new annotation will be added called `strimzi.io/mirrormaker-connector`.
+It will be required when `strimzi.io/connector-offsets` is set.
+The value will be a comma separated list of the names of the connectors to apply the action to, for example `east-kafka->west-kafka.MirrorSourceConnector,east-kafka->west-kafka.MirrorCheckpointConnector`.
+
+If the `strimzi.io/mirrormaker-connector` annotation is set, but the `strimzi.io/connector-offsets` annotation is missing the operator will take no action but add a warning message.
+
+### Listing offsets
+
+To list the current offsets the user will annotate the KafkaMirrorMaker2 resource with the new annotation `strimzi.io/connector-offsets` set to `list`, and the `strimzi.io/mirrormaker-connector` annotation.
+After the annotations are added, on the next reconciliation the operator will fetch the current offsets for the connector(s) and create Kubernetes ConfigMap(s) containing the JSON response(s).
+The operator will add an owner reference to the Kubernetes ConfigMap(s) pointing to the KafkaMirrorMaker2 resource that was annotated.
+If the ConfigMap(s) already exist the operator will patch the ConfigMap(s) with updated data and not add an owner reference.
+This means any existing keys in the ConfigMap(s) that the operator is not updating will remain as before.
+
+Since the operator reconciles connectors one-by-one, after each list operation the operator will update the `strimzi.io/mirrormaker-connector` annotation to remove the connector it has just listed offsets for.
+Once all list operations are complete the operator will then remove both the `strimzi.io/mirrormaker-connector` and `strimzi.io/connector-offsets` annotations from the KafkaMirrorMaker2 CR.
+This will make it easier to recover from errors.
```

Comment:
They should be all listed within the same reconciliation. So why do they need to be updated one by one? Why not update them just once at the end or keep it if something fails.

Reply:
Yeah, but ...
* You do also a lot of patching which can cause issues and rewrite other annotations
* Wouldn't the user want to reset them at the same time and if one fails reset them all again when working with multiple connectors?

I do not have completely strong feelings for this if you wanna keep it. But if I was writing this, I would do all at once at the end.

PS: How do we do it with task restarts? Isn't that similar?

Reply:
For restarting connectors/tasks, you can only restart a single connector/task at once, so the annotation is `"strimzi.io/restart-connector-task=<mirrormaker_connector_name>:<task_id>"`